rfcs/0027-frontier-synchronization.md

```diff
@@ -0,0 +1,188 @@
+
+#### Detached Mask State Management
+
+In the current system, when a mask is destroyed, all of it's children masks are also destroyed. Masks will now additionally provide the ability to attach "destruction hooks". This will be used to cleanup full breadcrumbs and downgrade them to partial breadcrumbs if the breadcrumb's parent mask chain becomes inaccessible. This allows code outside of synchronized frontier reads to continue to query breadcrumbs from the frontier and build successor breadcrumbs off of them safely. This is important in the context of catchup since new subtrees of breadcrumbs are built asynchronously to the frontier being updated. If a subtree is built off of a breadcrumb which is later removed from the frontier (thus having it's mask destroyed), then that subtree will subsequently be marked as partial and any masks in it will be destroyed and inaccessible.
+
+IMPLEMENTATION NOTE: The entire chain of mask destruction hooks needs to be executed synchronously within a single async cycle in order to update breadcrumb subtrees without fear of race conditions.
+
+#### Frontier Read Monad
+
+A new monad is introduced for specifying reads from the transition frontier (see [alternatives section](#alternatives-misread) for explanation on why). All existing read functions on the transition frontier will instead by turned into functions which do not take in a transition frontier and return a result wrapped in the `'a Transition_frontier.Read.t` monad. The monad is designed to interact with deferreds so that async programming is still accessible during reads when necessary.
+
+#### New Frontier Interface
+
+Below is a partial signature for what the new transition frontier interface would look like. Note that all base read/write functions from the transition frontier are removed, and instead, only the read monad and write actions are accessible. The only way to interact with the transition frontier thus becomes the `read` and `write` functions.
+
+```ocaml
+module Transition_frontier : sig
+  module Breadcrumb : sig
+    module Full : sig
+      type t
+      val block : t -> Block.Validated.t
+      val staged_ledger : t -> Transaction_snark_scan_state.t
+    end
+
+    module Partial : sig
+      type t
+      val block : t -> Block.Validated.t
+      val scan_state : t -> Transaction_snark_scan_state.t
+    end
+
+    type full
+    type partial
+
+    type 'typ t =
+      | Full : Full.t -> full t
+      | Partial : Partial.t -> partial t
+    type 'typ breadcrumb = 'typ t
+
+    val external_transition : t -> External_transition.Validated.t
+
+    module Any : sig
+      type t = T : _ breadcrumb -> t
+      type any = t
+
+      val wrap : _ breadcrumb -> t
+
+      module Ref : sig
+        (* intentionally left abstract to enforce synchronous access *)
+        type t
+
+        val wrap : any -> t
+        (* downgrades a full to a partial; returns error if already a partial *)
+        val downgrade : t -> unit Or_error.t
```

Comment:
Can this return a `t -> Partial.t` instead? Or if you want to keep the signaling of "this is already a partial":

```ocaml
t -> Partial.t * [`Already_partial, `Downgraded]
```

Reply:
It needs to mutate the underlying reference, hence why it can cause an error. You can ignore the error if you want. I'm not sure if I want to expose the partial directly after a downgrade or not, but I will think about it some more.